`config`: rework `delete.retention.ms` validation and add `tx_retention_ms` [SC&R #3]

WillemKauf opened the pull request:
This PR is third in the series of PRs for the safe compaction & recovery [SC&R] project.

This PR reworks validation of the topic level property `delete.retention.ms` as well as the cluster level property `tombstone_retention_ms` to remove the previous requirement that tiered storage be disabled in order to set their values.

The default value for `delete.retention.ms` at the topic level is still `DISABLED` for backwards compatibility reasons and `tombstone_retention_ms` is still left as `std::nullopt` after these changes. A sane default/enablement of these features may come in a later PR.

The major reason for this is that `delete.retention.ms` will ALSO control the removal of transactional batches during compaction (as a side note, this now makes the cluster config `tombstone_retention_ms` a bit of a misnomer- however, we couldn't re-use `delete_retention_ms` at the cluster level at the time of addition due to other historical config reasons within `redpanda`). 

This PR also adds the property `tx_retention_ms` to `storage::compaction_config` and `storage::ntp_config`.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

WillemKauf committed:
`storage`: add `delete_retention_ms` validation tests

These tests ensure that `ntp_config::tombstone_retention_ms()`
and `ntp_config::tx_retention_ms()` return expected values when tiered
storage is enabled/disabled, that is,

* TS enabled -> `tombstone_retention_ms() == std::nullopt`,
                `tx_retention_ms() == value`

* TS disabled -> `tombstone_retention_ms() == value`,
                 `tx_retention_ms() == value`

WillemKauf committed:
`storage`: add `ntp_config::tx_retention_ms()`

This commit adds an accessor function to `ntp_config` for
`tx_retention_ms`.

It also slightly reworks the logic in `ntp_config::tombstone_retention_ms()`
to make use of a new, generic `ntp_config::delete_retention_ms()`.

The checks in `tombstone_retention_ms()` and `tx_retention_ms()` differ
because we need to ensure that the tombstone retention time is always
`std::nullopt` when tiered storage is enabled (as there still exists
race conditions with tombstone removal and tiered storage uploads),
whereas `tx_retention_ms()` does not have the same restriction (we always
upload below LSO, and the removal of local transactional batches does not
impact the correctness of the log if the same transactional batches
still exist in the cloud).

WillemKauf committed:
`storage`: add `compaction_config::tx_retention_ms`

Further plumbing of the `ntp_config::tx_retention_ms` down
into the storage layer.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68330-->
<!--ad78f957-1d9d-4c81-93cb-fa4253ee6930-->
<!--ac2b7b4d-c062-4790-bd67-856696a22b1c-->
<!--title-->
## Retry command for [Build#68330](https://buildkite.com/redpanda/redpanda/builds/68330)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"sliding_window","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"sliding_window","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--edb4149e-aad5-4d58-8e3c-2dc5e53da30b-->
<!--TEST RESULTS-->
<!--d9c4436a-fca9-4dc9-8685-e79e60e47307-->
<!--TEST RESULTS-->
<!--70ca5cd4-1201-46f3-826d-e32cf51e24fd-->
<!--TEST RESULTS-->
<!--bb3f8fbb-bcfb-4838-9045-4ec676236645-->
<!--TEST RESULTS-->
<!--f4f81ff3-5e6d-4d19-82a9-d357cde3e273-->
<!--TEST RESULTS-->
<!--0e8cf3f1-6cae-4426-8b81-32a66639c08a-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68330">build#68330</a></summary>


| test_class                | test_method                             | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:----------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ArchivalTest              | test_connection_flicker                 | {"cloud_storage_type": 2}                                                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68330#0197cbbf-302b-44e8-8d1c-69c11ce2b9aa | FLAKY         | 20/21    | upstream reliability is '98.98477157360406'. current run reliability is '95.23809523809523'. drift is 3.74668 and the allowed drift is set to 50. The test should PASS |
| ClusterConfigTest         | test_valid_settings                     |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68330#0197cc3f-f18d-4e64-a66b-c11582ac3bd3 | FLAKY         | 15/21    |                                                                                                                                                                        |
| RandomNodeOperationsTest  | test_node_operations                    | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68330#0197cc3f-f18d-4e64-a66b-c11582ac3bd3 | FLAKY         | 3/6      |                                                                                                                                                                        |
| RandomNodeOperationsTest  | test_node_operations                    | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68330#0197cbc0-f403-45d6-bbb7-ae1c5aeca0fe | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest  | test_node_operations                    | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68330#0197cc3f-f1fa-42ba-a9da-1de2d8364981 | FLAKY         | 1/4      |                                                                                                                                                                        |
| RandomNodeOperationsTest  | test_node_operations                    | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68330#0197cbc0-f403-45d6-bbb7-ae1c5aeca0fe | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest  | test_node_operations                    | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68330#0197cc3f-f1fa-42ba-a9da-1de2d8364981 | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest  | test_node_operations                    | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68330#0197cbc0-f405-4f22-9741-ff8448e0b893 | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest  | test_node_operations                    | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68330#0197cc2c-bbff-4a33-b6f0-a957a154e63f | FLAKY         | 10/20    |                                                                                                                                                                        |
| RandomNodeOperationsTest  | test_node_operations                    | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68330#0197cbc0-f405-447f-92cd-ab743a79478a | FLAKY         | 8/18     |                                                                                                                                                                        |
| TestTieredStoragePause    | test_safe_pause_resume                  | {"allow_gaps_cluster_level": false, "allow_gaps_topic_level": true}                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68330#0197cc2c-bbff-4a33-b6f0-a957a154e63f | FLAKY         | 19/21    |                                                                                                                                                                        |
| RecreateTopicMetadataTest | test_recreated_topic_metadata_are_valid | {"replication_factor": 3}                                                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68330#0197cbc0-f406-46f4-b7dd-8d0811553d69 | FLAKY         | 19/21    | upstream reliability is '97.42647058823529'. current run reliability is '90.47619047619048'. drift is 6.95028 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68373">build#68373</a></summary>


| test_class                   | test_method                                         | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------|:----------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ArchivalTest                 | test_retention_archival_coordination                | {"acks": -1, "cloud_storage_type": 2}                                                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68373#0197ce7a-ded1-4c1a-8c66-04a393efca97 | FAIL          | 0/1      |                                                                                                                                                                        |
| CloudStorageChunkReadTest    | test_read_when_segment_size_smaller_than_chunk_size |                                                                                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68373#0197ce79-b57a-4836-965e-23bc210f74d9 | FLAKY         | 20/21    | upstream reliability is '98.53658536585365'. current run reliability is '95.23809523809523'. drift is 3.29849 and the allowed drift is set to 50. The test should PASS |
| CloudStorageTimingStressTest | test_cloud_storage_with_partition_moves             | {"cleanup_policy": "compact,delete"}                                                                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68373#0197ce7a-ded5-4d4c-9210-5324e9e41f87 | FLAKY         | 20/21    | upstream reliability is '97.08520179372198'. current run reliability is '95.23809523809523'. drift is 1.84711 and the allowed drift is set to 50. The test should PASS |
| DeleteRetentionMsTest        | test_alter_topic_config                             |                                                                                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68373#0197ce79-b57a-4836-965e-23bc210f74d9 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DeleteRetentionMsTest        | test_alter_topic_config                             |                                                                                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68373#0197ce7a-ded4-4980-847d-56e61f6128c6 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DeleteRetentionMsTest        | test_get_cluster_config                             |                                                                                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68373#0197ce79-b57b-4de8-bdd4-44d3c08ab689 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DeleteRetentionMsTest        | test_get_cluster_config                             |                                                                                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68373#0197ce7a-ded5-4d4c-9210-5324e9e41f87 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DescribeTopicsTest           | test_describe_topics_with_documentation_and_types   |                                                                                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68373#0197ce79-b57e-44fa-a71c-3a64e4d7dc56 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DescribeTopicsTest           | test_describe_topics_with_documentation_and_types   |                                                                                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68373#0197ce7a-ded2-4dc0-be2c-c6c202c1f73c | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest     | test_node_operations                                | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68373#0197ce7a-ded5-48b7-8130-eaae6e82f9cc | FLAKY         | 12/21    | upstream reliability is '100.0'. current run reliability is '57.14285714285714'. drift is 42.85714 and the allowed drift is set to 50. The test should PASS            |
| RandomNodeOperationsTest     | test_node_operations                                | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68373#0197ce7a-ded8-4590-afb1-a738629da63e | FLAKY         | 2/5      |                                                                                                                                                                        |
| RandomNodeOperationsTest     | test_node_operations                                | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68373#0197ce7a-ded2-4dc0-be2c-c6c202c1f73c | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest     | test_node_operations                                | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68373#0197ce7a-ded3-4b9c-81ff-edfb0bb17f29 | FLAKY         | 8/20     |                                                                                                                                                                        |
| RandomNodeOperationsTest     | test_node_operations                                | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68373#0197ce7a-ded4-4790-9b1f-842d6e280688 | FLAKY         | 9/17     |                                                                                                                                                                        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68414">build#68414</a></summary>


| test_class                          | test_method                             | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:------------------------------------|:----------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| archival_metadata_stm_gtest_fixture | test_same_term_sync_pending_replication |                                                                                                                                                 | unit        | https://buildkite.com/redpanda/redpanda/builds/68414#0197d0ac-276b-4938-8b71-6f36e6cfdcbf | FAIL          | 0/1      |                                                                                                                                                                        |
| ShadowIndexingWhileBusyTest         | test_create_or_delete_topics_while_busy | {"cloud_storage_type": 2, "short_retention": true}                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68414#0197d0db-2725-448e-9855-5b1fdb6da50c | FLAKY         | 15/19    |                                                                                                                                                                        |
| RandomNodeOperationsTest            | test_node_operations                    | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68414#0197d0db-2728-4d05-83ad-3e6cc4ab97cb | FLAKY         | 2/5      |                                                                                                                                                                        |
| RandomNodeOperationsTest            | test_node_operations                    | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68414#0197d0db-2729-494a-a8fa-51dc9f738de6 | FAIL          | 0/4      |                                                                                                                                                                        |
| RandomNodeOperationsTest            | test_node_operations                    | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68414#0197d0db-2724-4464-bb98-921acd0e3ed4 | FLAKY         | 3/6      |                                                                                                                                                                        |
| RandomNodeOperationsTest            | test_node_operations                    | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68414#0197d0db-2725-44c6-9e5d-7b4a319ba6d8 | FLAKY         | 3/6      |                                                                                                                                                                        |
| RandomNodeOperationsTest            | test_node_operations                    | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68414#0197d0db-2724-4464-bb98-921acd0e3ed4 | FLAKY         | 2/6      |                                                                                                                                                                        |
| TopicDeleteCloudStorageTest         | drop_lifecycle_marker_test              | {"cloud_storage_type": 2}                                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68414#0197d0ef-36d8-439c-b667-ed6dce67a840 | FLAKY         | 19/21    | upstream reliability is '97.21669980119285'. current run reliability is '90.47619047619048'. drift is 6.74051 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68447">build#68447</a></summary>


| test_class                  | test_method                | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------------|:---------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest    | test_node_operations       | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68447#0197d1e3-547d-4d1a-9b22-329de31bc5d7 | FLAKY         | 1/3      |                                                                                                                                                                         |
| RandomNodeOperationsTest    | test_node_operations       | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68447#0197d1e3-547f-47be-aa2b-4184a0099911 | FAIL          | 0/1      |                                                                                                                                                                         |
| RandomNodeOperationsTest    | test_node_operations       | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68447#0197d1e3-5480-4d4b-bbe6-3ae94cb5d6fd | FLAKY         | 3/9      |                                                                                                                                                                         |
| RandomNodeOperationsTest    | test_node_operations       | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": false, "with_iceberg": false}         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68447#0197d1e3-547e-405a-95d9-c5d9d7005a97 | FLAKY         | 20/21    | upstream reliability is '97.91666666666666'. current run reliability is '95.23809523809523'. drift is 2.67857 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest    | test_node_operations       | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68447#0197d1e3-5480-4d4b-bbe6-3ae94cb5d6fd | FLAKY         | 3/8      |                                                                                                                                                                         |
| RandomNodeOperationsTest    | test_node_operations       | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68447#0197d1e3-5482-47e8-a713-2fdcb475b082 | FLAKY         | 7/21     | upstream reliability is '99.19678714859438'. current run reliability is '33.33333333333333'. drift is 65.86345 and the allowed drift is set to 50. The test should FAIL |
| RandomNodeOperationsTest    | test_node_operations       | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68447#0197d1e3-5484-494c-9737-22088c30d3bf | FLAKY         | 7/21     | upstream reliability is '100.0'. current run reliability is '33.33333333333333'. drift is 66.66667 and the allowed drift is set to 50. The test should FAIL             |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 2}                                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68447#0197d1e8-7bbd-4eeb-a158-538c8f4b952a | FLAKY         | 20/21    | upstream reliability is '97.24950884086444'. current run reliability is '95.23809523809523'. drift is 2.01141 and the allowed drift is set to 50. The test should PASS  |
| TopicRecoveryTest           | test_no_data               | {"cloud_storage_type": 2}                                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68447#0197d1e3-547f-47be-aa2b-4184a0099911 | FAIL          | 0/1      |                                                                                                                                                                         |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68490">build#68490</a></summary>


| test_class                  | test_method                | test_arguments                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------------|:---------------------------|:----------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerGroupTest           | test_group_lag_metrics     |                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68490#0197d686-c137-4149-91ba-fb812c3dea53 | FLAKY         | 20/21    | upstream reliability is '99.71751412429379'. current run reliability is '95.23809523809523'. drift is 4.47942 and the allowed drift is set to 50. The test should PASS  |
| DatalakeClusterRestoreTest  | test_basic                 | {"catalog_type": "nessie", "cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68490#0197d687-19de-4f36-8d6c-1dbe5726f6b5 | FLAKY         | 20/21    | upstream reliability is '99.71671388101983'. current run reliability is '95.23809523809523'. drift is 4.47862 and the allowed drift is set to 50. The test should PASS  |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 2}                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68490#0197d687-19df-4bf4-84cd-e00deb3bc3bc | FLAKY         | 18/21    | upstream reliability is '97.18045112781954'. current run reliability is '85.71428571428571'. drift is 11.46617 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68496">build#68496</a></summary>


| test_class                   | test_method                        | test_arguments                                                                                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------|:-----------------------------------|:-------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageTimingStressTest | test_cloud_storage                 | {"cleanup_policy": "compact,delete"}                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68496#0197dabb-06d3-489d-b3cf-209be7142ef4 | FLAKY         | 19/21    | upstream reliability is '96.06060606060606'. current run reliability is '90.47619047619048'. drift is 5.58442 and the allowed drift is set to 50. The test should PASS |
| DataMigrationsApiTest        | test_migrated_topic_data_integrity | {"params": {"cancellation": {"dir": "in", "stage": "executing"}, "use_alias": false}, "transfer_leadership": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68496#0197dac1-b80e-455d-ac62-7aabeb4a494a | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<!--/body-->

WillemKauf commented:
Lot of `RandomNodeOperationTest` failures with mixed_versions- I think I need to feature gate the `tombstone_retention_ms` default changes.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68373-->
<!--7b1d6bf8-e9d6-4af9-a329-9ff7179b1ae3-->
<!--eb02e0a4-1d38-4c3e-a85c-ce030c36ea8d-->
<!--c84f68b0-1448-4485-b485-85a3e4c35d4f-->
<!--288d73d2-4ad7-4593-b590-899d480a4948-->
<!--d72240d5-2ded-4bdf-ba13-4e1cda4c5b16-->
<!--title-->
## Retry command for [Build#68373](https://buildkite.com/redpanda/redpanda/builds/68373)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/delete_retention_ms_test.py::DeleteRetentionMsTest.test_get_cluster_config
tests/rptest/tests/describe_topics_test.py::DescribeTopicsTest.test_describe_topics_with_documentation_and_types
tests/rptest/tests/delete_retention_ms_test.py::DeleteRetentionMsTest.test_alter_topic_config
```
<!--/body-->

WillemKauf commented:
Removing change to default value of `tombstone_retention_ms` from `std::nullopt` to `24h` for now. Not sure why the mixed version tests are failing, but to be investigated.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68447-->
<!--06e8c2c5-cd7b-4933-98e4-ba36d08ba9db-->
<!--d8b156ef-654a-4eae-aaa8-e581a112c5e4-->
<!--title-->
## Retry command for [Build#68447](https://buildkite.com/redpanda/redpanda/builds/68447)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"sliding_window","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"chunked_sliding_window","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
```
<!--/body-->

WillemKauf committed:
`storage`: add more output for `ntp_config::default_override`

Some variables which were not being captured in the `ostream` output
for the `default_override` object are added back here.

WillemKauf commented:
Yeah. I'm going to have to feature gate the tombstone config validation changes- mixed version cluster will throw bad errors if a `create_topics` request is dispatched to a previous version node with more stringent conditions on `tombstone_retention_ms` validation...

WillemKauf committed:
`config`: rework `delete.retention.ms` validation

This commit reworks validation of the topic level property `delete.retention.ms`
as well as the cluster level property `tombstone_retention_ms` to remove the
previous requirement that tiered storage be disabled in order to set their
values.

The major reason for this is that `delete.retention.ms` will ALSO control
the removal of transactional batches during compaction (as a side note, this
now makes the cluster config `tombstone_retention_ms` a bit of a misnomer-
however, we couldn't re-use `delete_retention_ms` at the cluster level
at the time of addition due to other historical config reasons within
`redpanda`).

Probing the `ntp_config()` for a topic's `tombstone.retention.ms` should
still return `std::nullopt` in the case of a tiered storage enabled topic
after these changes.

WillemKauf committed:
`storage`: rename `ntp_config::tombstone_retention_ms`

After using this property to also set the transactional batch delete
horizon, `tombstone_retention_ms` will be somewhat of a under-descriptive
variable name.

Rename it to `delete_retention_ms`.

andrwng commented:
Tagging Bharath for a look at the tx related config

andrwng commented on src/v/storage/ntp_config.h:
```diff
@@ -316,22 +316,14 @@ class ntp_config {
     }
 
-    std::optional<std::chrono::milliseconds> tombstone_retention_ms() const {
+    std::optional<std::chrono::milliseconds> delete_retention_ms() const {
```
Should callers ever use this? Can it be made private?

WillemKauf replied:
>Should callers ever use this? Can it be made private? 

I use it in a future commit. I'll leave it public for now, if I can rework the future commit with it private I'll make the change then if that is okay.
